### PR TITLE
Kokkos: Don't let Kokkos set CMAKE_CXX_FLAGS

### PR DESCRIPTION
**Pull request by masterleinad**
@trilinos/kokkos

## Motivation
`Kokkos` is currently setting `CMAKE_CXX_FLAGS` for `Trilinos` which is problematic.
- the errors in https://github.com/trilinos/Trilinos/issues/12420 demonstrate that this breaks linking for anything that doesn't also use `target_link_libraries(target Kokkos::kokkos)` (or an alias for `Kokkos::kokkos`) since the compile option `fsycl-targets=spir64_gen` can only be used if `-Xsycl-target-backend` is used as linker flag.
- `CMAKE_CXX_FLAGS` forces compile options on all targets that don't need it overriding possibly more suitable flags.

The lines modified here were last touched in https://github.com/kokkos/kokkos/pull/2660 suggesting that `Trilinos` was relying on `CMAKE_CXX_FLAGS` being set by `Kokkos` at least at some point. I assume this has become obsolete by using `target_link_libraries` more consistently. @bartlettroscoe does that match your expectations?
Note that the changes here are also included in https://github.com/trilinos/Trilinos/pull/11779 which makes `Kokkos` a `TriBits`-compatible raw `CMake` project snapshotted in `Trilinos`.

## Related Issues
 
* Fixes https://github.com/trilinos/Trilinos/issues/12420

## Testing
Compiling `Trilinos` with `Kokkos+SYCL`.

### CODE REVIEW

**Review by lucbv — APPROVED** [review]

**Review by bartlettroscoe — CHANGES_REQUESTED** [review]
@maartenarnst,

I agree, we should not be expecting `CMAKE_CXX_FLAGS` to be passed along to downstream CMake customers who pull in Kokkos (or any Trilinos package) with `find_package(<upstreamPkg> ...)`.  Instead, critical compiler options should be propagated using the `INTEFACE_COMPILE_OPTIONS` target property using `target_link_libraries(<dowsntreamTarget> <INTERFACE|PUBLIC|PRIVIATE> <upstreamTarget>)`.

But then why not also remove setting `CMAKE_CXX_FLAGS` as a local var which will hide `CMAKE_CXX_FLAGS` set as a cache var and not allow the user to tweak any of these flags?  I think this should be addressed before merging this PR (see below).

File: packages/kokkos/CMakeLists.txt
Comment (masterleinad) [reply packages/kokkos/CMakeLists.txt:294]: This code path is only used for `Trilinos` anyway and I tried to be conservative. Hopefully, we can move forward with https://github.com/kokkos/kokkos/pull/6164 soon and clean up the interaction between `Trilinos` and `Kokkos` some more.

I'm happy to give it a shot to remove the whole block and all related code already.

Comment (bartlettroscoe) [reply packages/kokkos/CMakeLists.txt]: @masterleinad, if you want to defer final cleanup, then that is fine and we can go ahead and merge this.  But given problems with the PR builds (see #12571), there is no hurry.  I will go ahead and remove my "request changes" and you can decide what you want to do with this.

**Review by bartlettroscoe — COMMENTED** [review]
This cleanup is not critical and since there is a more significant refactoring of the Kokkos build under Trilinos coming (see https://github.com/kokkos/kokkos/pull/6164), I will defer to others if to address the last CMAKE_CXX_FLAGS issue in this PR or not.

**Comment by bartlettroscoe** [comment]
FYI: Don't bother putting on `AT: RETEST` until:

* #12571 

is resolved.

**Comment by masterleinad** [comment]
Can we retest this and come to a conclusion if we want both commits or only the first one? It seems this is also causing problems for `HIP`, see https://github.com/kokkos/kokkos/issues/6736.

**Comment by bartlettroscoe** [comment]
> Can we retest this and come to a conclusion if we want both commits or only the first one? It seems this is also causing problems for `HIP`, see [kokkos/kokkos#6736](https://github.com/kokkos/kokkos/issues/6736).

@masterleinad, let's pull the trigger on the suggestions in [above](https://github.com/trilinos/Trilinos/pull/12572#pullrequestreview-1760143361) in a new commit and then trigger the retest?

**Comment by masterleinad** [comment]
> @masterleinad, let's pull the trigger on the suggestions in https://github.com/trilinos/Trilinos/pull/12572#pullrequestreview-1760143361 in a new commit and then trigger the retest?

@bartlettroscoe I pushed a more conservative version than I had before now that avoids setting `CMAKE_CXX_FLAGS` and deletes all related (now unused) variables.

**Comment by ndellingwood** [comment]
@masterleinad we'll need these changes applied to the kokkos repo, as well as applied to the release-candidate-4.2.01 branch if this PR makes it into Trilinos before the patch release (to avoid clobber)

**Review by bartlettroscoe — DISMISSED** [review]
Looks good to me.  Rip it out.  Less is more :-)

File: packages/kokkos/CMakeLists.txt
Comment (maartenarnst) [inline packages/kokkos/CMakeLists.txt:288]: Removing this block of code seems to have the side effect that the cuda fe flags are no longer passed to Trilinos through KOKKOS_ALL_COMPILE_OPTIONS. May be good to doublecheck if that’s intended.

Comment (masterleinad) [reply packages/kokkos/CMakeLists.txt:288]: You're right. For now, it's better not to modify `KOKKOS_ALL_COMPILE_OPTIONS`. I pushed a corresponding commit.

**Comment by masterleinad** [comment]
> @masterleinad we'll need these changes applied to the kokkos repo, as well as applied to the release-candidate-4.2.01 branch if this PR makes it into Trilinos before the patch release (to avoid clobber)

I agree. I just wanted to run the pull request first through `Trilinos` testing since I won't affect a standalone `Kokkos` installation at all.

File: packages/kokkos/CMakeLists.txt
Comment (maartenarnst) [reply packages/kokkos/CMakeLists.txt:288]: OK!

**Review by maartenarnst — APPROVED** [review]
Thanks for these changes!

**Comment by maartenarnst** [comment]
I confirm that these changes solve the issue reported in 
- kokkos/kokkos#6736

i.e., with these changes, the way Kokkos sets the `-x hip` flag works when using the Cray compiler, both in the case of compiling Kokkos as a standalone and in the case of compiling Kokkos as part of Trilinos.

File: packages/kokkos/CMakeLists.txt
Comment (bartlettroscoe) [reply packages/kokkos/CMakeLists.txt:288]: The handling of `KOKKOS_ALL_COMPILE_OPTIONS` is a bit confusing to me so I don't understand the logic here.  From what I can tell, these options are written to a configured file  `KokkosTrilinosConfig.cmake` and added to the `Kokkos::kokkos` target using `target_compiler_options()`.  Then that file gets read in and written to the end of the generated file `KokkosConfig.cmake` which gets installed.  Why not just directly add those compile options to the individual Kokkos libraries using `target_compiler_options()`?  Then those options would be carried along with the direct Kokkos libraries no matter how they are used.
s

Comment (masterleinad) [reply packages/kokkos/CMakeLists.txt:288]: I agree but `KOKKOS_ALL_COMPILE_OPTIONS` doesn't cause the issue we are trying to fix here and we wanted to keep the changes here minimal.
Note that all this logic is just for `Trilinos`. A standalone `Kokkos` doesn't even go through these lines of code.

With the refactoring in #11779 we should eliminate all the weirdness we had in Kokkos just to make Trilinos work.

**Review by crtrott — DISMISSED** [review]

**Comment by masterleinad** [comment]
What do the test failures look like?

**Comment by bartlettroscoe** [comment]
> What do the test failures look like?

@masterleinad, the main error was building the object file:

* `packages/kokkos/simd/unit_tests/CMakeFiles/Kokkos_UnitTest_SIMD.dir/TestSIMD.cpp.o`

which was:

```
In file included from ../Trilinos/packages/kokkos/simd[Kokkos_SIMD.hpp:29](https://github.com/trilinos/Trilinos//blob/master/Kokkos_SIMD.hpp#L29),
                 from ../Trilinos/packages/kokkos/simd/unit_tests/include/TestSIMD_MathOps.hpp:20,
                 from ../Trilinos/packages/kokkos/simd/unit_tests/TestSIMD.cpp:17:
../Trilinos/packages/kokkos/simd[Kokkos_SIMD_Scalar.hpp:227](https://github.com/trilinos/Trilinos//blob/master/Kokkos_SIMD_Scalar.hpp#L227):2: error: extra ';' [-Werror=pedantic]
 };
  ^
../Trilinos/packages/kokkos/simd[Kokkos_SIMD_Scalar.hpp:235](https://github.com/trilinos/Trilinos//blob/master/Kokkos_SIMD_Scalar.hpp#L235):2: error: extra ';' [-Werror=pedantic]
 };
  ^
../Trilinos/packages/kokkos/simd[Kokkos_SIMD_Scalar.hpp:243](https://github.com/trilinos/Trilinos//blob/master/Kokkos_SIMD_Scalar.hpp#L243):2: error: extra ';' [-Werror=pedantic]
 };
  ^
../Trilinos/packages/kokkos/simd[Kokkos_SIMD_Scalar.hpp:251](https://github.com/trilinos/Trilinos//blob/master/Kokkos_SIMD_Scalar.hpp#L251):2: error: extra ';' [-Werror=pedantic]
 };
  ^
[CTest: warning matched] cc1plus: all warnings being treated as errors
```

The test [Phalanx_experimental_mdfield_refactor_MPI_1](https://trilinos-cdash.sandia.gov/test/45623445) also segfaulted in the build [PR-12572-test-rhel7_sems-intel-2021.3-sems-openmpi-4.0.5_release-debug_shared_no-kokkos-arch_no-asan_no-complex_fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-off_no-package-enables-1708 ](https://trilinos-cdash.sandia.gov/build/1360183)[(ascic114)](https://trilinos-cdash.sandia.gov/viewSite.php?siteid=53) but I would guess that is unrelated.

**Comment by dalg24** [comment]
These were fixed in https://github.com/kokkos/kokkos/pull/6510

**Comment by masterleinad** [comment]
I'm surprised that these problems only show up in this pull request. Are pull requests tested rebased on top of `develop`?
Otherwise, I can rebase manually and hope that that fixes it.

**Comment by masterleinad** [comment]
I rebased and cherry-picked kokkos/kokkos#6510 so we can retest. The best solution would be to merge this as part of the upcoming Kokkos patch release anyway.

**Comment by masterleinad** [comment]
What's the failure this time?

**Comment by ndellingwood** [comment]
@masterleinad looks like some of the jobs did not complete, related to https://github.com/trilinos/Trilinos/issues/12696 and possibly backlog of jobs waiting in the queue when the GPU nodes came back online. Retesting

**Comment by ndellingwood** [comment]
Looks like the `sems-gnu-8.3.0-openmpi` test failed due to running out of space
```
/usr/bin/ld: final link failed: No space left on device
```

No failures reported in the other builds

**Comment by masterleinad** [comment]
We merged this one on the `Kokkos`-side and will pull it in with the patch release, see https://github.com/kokkos/kokkos/pull/6756.

**Pull request closed without merging**
